Doc/library/importlib.rst

```diff
@@ -1284,66 +1288,71 @@ find and load modules.
    A specification for a module's import-system-related state.  This is
    typically exposed as the module's ``__spec__`` attribute.  In the
    descriptions below, the names in parentheses give the corresponding
-   attribute available directly on the module object.
-   E.g. ``module.__spec__.origin == module.__file__``.  Note however that
-   while the *values* are usually equivalent, they can differ since there is
-   no synchronization between the two objects.  Thus it is possible to update
-   the module's ``__path__`` at runtime, and this will not be automatically
-   reflected in ``__spec__.submodule_search_locations``.
+   attribute available directly on the module object. For example,
+   ``module.__spec__.name == module.__name__``.  Note however that while
+   the *values* are usually equivalent, they can differ since there is no
+   synchronization between the two objects.  For example, it is possible
+   to update the module's ``__name__`` at runtime and this will not be
+   automatically reflected in the module's ``__spec__.name``, or the
+   other way round.
 
    .. versionadded:: 3.4
 
    .. attribute:: name
 
    (``__name__``)
 
-   A string for the fully-qualified name of the module.
+   The module's fully-qualified name.
 
    .. attribute:: loader
 
    (``__loader__``)
 
-   The loader to use for loading.  For namespace packages this should be
-   set to ``None``.
+   The module's loader.
 
    .. attribute:: origin
 
    (``__file__``)
 
-   Name of the place from which the module is loaded, e.g. "builtin" for
-   built-in modules and the filename for modules loaded from source.
-   Normally "origin" should be set, but it may be ``None`` (the default)
-   which indicates it is unspecified (e.g. for namespace packages).
+   The origin of the module's source code.
```

Comment:
The original text was very careful to be precise.  "origin" is more than just for source code.  It is exactly "Name of the place from which the module is loaded".  Please leave that phrasing.

Note that the only purpose of `ModuleSpec` is for a finder to provide all necessary information the loader needs to load the module.  Especially in the case of "origin", the value (and meaning) of each spec attr is up to the finder.  The finder is then responsible for making sure that the loader understands the spec.

So be careful not to make it seem like we are defining all possible values of "origin" here.  We can't because the set of possible values is effectively infinite (even if in practice there won't be many unknown ones).  One or two *examples* are enough to help readers understand.  I can see how listing all the default ones would be useful, but the text should be clear that it could be anything.

Also, of all the default finders and loaders the only one that really matters here is for files (.py, .pyc, .so).  Of all the examples that's the one that should *not* be removed.

Finally, the original last sentence is also important and should be left alone (or at most cleaned up a little).

Reply:
Okay, let's keep the original definition and "e.g." to emphasize that where are not exhaustive. What about:

> The name of the place from which the module was loaded.
> E.g., it is ``'built-in'`` for built-in modules, ``'frozen'`` for frozen modules, the relative path for modules run as a script, the absolute path for modules run as a module or imported modules and ``'<stdin>'`` for modules run from standard input. It is ``None`` (the default) when unspecified, e.g., for namespace packages.

Reply:
re-visited in a follow-up review